bam-index/src/main/java/org/victorchang/QnameSearcher.java

```diff
@@ -29,83 +33,103 @@ public QnameSearcher(KeyPointerReader keyPointerReader, BamRecordReader recordRe
-        FileChannel channelLevel0 = FileChannel.open(pathLevel0, READ);
-        long coffset = PointerPacker.INSTANCE.unpackCompressedOffset(start.getPointer());
-        int uoffset = PointerPacker.INSTANCE.unpackUnCompressedOffset(start.getPointer());
+        return pointersForQname.size();
+    }
 
-        if (coffset >= channelLevel0.size()) {
-            return 0;
-        }
-        channelLevel0.position(coffset);
-        InputStream inputStreamLevel0 = Channels.newInputStream(channelLevel0);
-
-        int found = 0;
-        Iterable<KeyPointer> indexLevel0 =  () -> keyPointerReader.read(inputStreamLevel0, uoffset).iterator();
-        for (KeyPointer x : indexLevel0) {
-            if (Arrays.equals(x.getKey(), input)) {
-                recordReader.read(bamFile, x.getPointer(), handler);
-                found++;
-            }
-            if (Arrays.compareUnsigned(x.getKey(), input) > 0) {
-                break;
-            }
-        }
+    List<Long> getPointersForQname(Path indexFolder, Set<String> qnames) throws IOException {
+        final Map<byte[], KeyPointer> qnameToPointer = getQnameToPointerMap(indexFolder, qnames);
 
-        inputStreamLevel0.close();
+        // Map from a pointer -> set of qnames in that level
+        final Map<Long, Set<byte[]>> pointerToQname = qnameToPointer.entrySet().stream()
+                .collect(toMap(e -> e.getValue().getPointer(), e -> Set.of(e.getKey()), this::concatSet));
 
-        return found;
+        return pointerToQname.entrySet().stream()
+                .flatMap(e -> getPointers(e.getValue(), indexFolder, e.getKey()).stream())
+                .collect(toList());
     }
 
-    public static class DebuggingHandler implements BamRecordHandler {
-
-        @Override
-        public void onHeader(SAMFileHeader header) {
+    /**
+     * Returns a map from the qname -> Key pointer location in the index file.
+     */
+    private Map<byte[], KeyPointer> getQnameToPointerMap(Path indexFolder, Set<String> qnames) throws IOException {
+        Path pathLevel1 = indexFolder.resolve("qname.1");
+        try (InputStream inputStreamLevel1 = Channels.newInputStream(FileChannel.open(pathLevel1, READ))) {
+            final List<KeyPointer> keyPointers = qnames.stream()
+                    .map(Ascii7Coder.INSTANCE::encode)
+                    .map(input -> new KeyPointer(0, input, input.length))
+                    .collect(toList());
+
+            final Map<byte[], KeyPointer> keyPointerMap = keyPointers.stream()
+                    .collect(toMap(KeyPointer::getKey, Function.identity()));
+
+            keyPointerReader.read(inputStreamLevel1)
+                    .forEach(indexPointer -> {
+                        for (KeyPointer keyPointer : keyPointers) {
+                            if (indexPointer.compareTo(keyPointer) < 0) {
+                                keyPointerMap.put(keyPointer.getKey(), indexPointer);
+                            }
```

Comment:
Similar to the other search, this could stop once it has found all the pointers that it needs, no? This currently goes through the whole index every time, or am I missing something?

Reply:
Yes, done. This can be further optimised by skipping already processed qnames, but it is probably not worth it.